File: packages/flutter/lib/src/http_gateway.dart
import 'dart:async';
import 'dart:convert';
import 'dart:io';
import 'gateway.dart';

/// Real gateway client over dart:io HttpClient.
class HttpGateway implements Gateway {
  final WaveConfig config;
  final HttpClient _client;

  HttpGateway(this.config, {HttpClient? client}) : _client = client ?? HttpClient();

  Future<Map<String, dynamic>> _post(String path, Map<String, dynamic> body, {String? bearer}) async {
    final req = await _client.postUrl(Uri.parse('${config.gatewayUrl}/$path'));
    req.headers.set('Content-Type', 'application/json');
    req.headers.set('apikey', config.anonKey);
    if (bearer != null) req.headers.set('Authorization', 'Bearer $bearer');
    req.add(utf8.encode(jsonEncode(body)));
    final res = await req.close();
    final text = await res.transform(utf8.decoder).join();
    final data = text.isEmpty ? <String, dynamic>{} : jsonDecode(text) as Map<String, dynamic>;
    if (res.statusCode >= 300) {
      throw Exception('gateway $path ${res.statusCode}: ${data['error'] ?? 'error'}');
    }
    return data;
  }

  @override
  Future<String> fetchToken() async {
    final data = await _post('partner-auth/token', {'key': config.publishableKey});
    final token = data['token'];
    if (token is! String) throw Exception('token missing');
    return token;
  }

  @override
  Future<Outcome> readOutcome(String token) async {
    final data = await _post('unlock-stream', {'card_id': config.userNumber}, bearer: token);
    final status = switch (data['status']) {
      'granted' => OutcomeStatus.granted,
      'denied' => OutcomeStatus.denied,
      _ => OutcomeStatus.pending,
    };
    return Outcome(status, data['reason'] as String?);
  }

  @override
  Future<Outcome> awaitOutcome(String token, Duration timeout, {Duration poll = const Duration(milliseconds: 500)}) async {
    final deadline = DateTime.now().add(timeout);
    while (DateTime.now().isBefore(deadline)) {
      try {
        final o = await readOutcome(token);
        if (o.status != OutcomeStatus.pending) return o;
      } catch (_) {}
      await Future.delayed(poll);
    }
    return const Outcome(OutcomeStatus.pending, null);
  }
}

File: packages/kotlin/src/main/kotlin/com/wave/unlock/HttpGateway.kt
package com.wave.unlock

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.delay
import kotlinx.coroutines.withContext
import java.net.URI
import java.net.http.HttpClient
import java.net.http.HttpRequest
import java.net.http.HttpResponse

/** Real gateway client over java.net.http. Minimal JSON parsing (no extra deps). */
class HttpGateway(
    private val config: WaveConfig,
    private val client: HttpClient = HttpClient.newHttpClient(),
) : Gateway {

    private fun post(path: String, body: String, bearer: String?): HttpResponse<String> {
        val builder = HttpRequest.newBuilder()
            .uri(URI.create("${config.gatewayUrl}/$path"))
            .header("Content-Type", "application/json")
            .header("apikey", config.anonKey)
        if (bearer != null) builder.header("Authorization", "Bearer $bearer")
        return client.send(builder.POST(HttpRequest.BodyPublishers.ofString(body)).build(),
            HttpResponse.BodyHandlers.ofString())
    }

    private fun field(json: String, key: String): String? {
        val m = Regex("\"$key\"\\s*:\\s*\"([^\"]*)\"").find(json)
        return m?.groupValues?.get(1)
    }

    override suspend fun fetchToken(): String = withContext(Dispatchers.IO) {
        val res = post("partner-auth/token", "{\"key\":\"${config.publishableKey}\"}", null)
        field(res.body(), "token") ?: throw IllegalStateException("token missing")
    }

    override suspend fun readOutcome(token: String): Outcome = withContext(Dispatchers.IO) {
        val res = post("unlock-stream", "{\"card_id\":\"${config.userNumber}\"}", token)
        val status = when (field(res.body(), "status")) {
            "granted" -> Outcome.Status.GRANTED
            "denied" -> Outcome.Status.DENIED
            else -> Outcome.Status.PENDING
        }
        Outcome(status, field(res.body(), "reason"))
    }

    override suspend fun awaitOutcome(token: String, timeoutMs: Long, pollMs: Long): Outcome {
        val deadline = System.currentTimeMillis() + timeoutMs
        while (System.currentTimeMillis() < deadline) {
            val o = runCatching { readOutcome(token) }.getOrNull()
            if (o != null && o.status != Outcome.Status.PENDING) return o
            delay(pollMs)
        }
        return Outcome(Outcome.Status.PENDING, null)
    }
}
